cloudevents-sdk-paho-mqtt/src/mqtt_consumer_record.rs

```diff
@@ -0,0 +1,239 @@
+
+        if let Some(hv) = self.headers.remove(headers::CONTENT_TYPE) {
+            visitor = visitor.set_attribute(
+                "datacontenttype",
+                MessageAttributeValue::String(String::from_utf8(hv).map_err(|e| {
+                    cloudevents::message::Error::Other {
+                        source: Box::new(e),
+                    }
+                })?),
+            )?
+        }
+
+        for (hn, hv) in self
+            .headers
+            .into_iter()
+            .filter(|(hn, _)| headers::SPEC_VERSION_HEADER != *hn && hn.starts_with("ce_"))
+        {
+            let name = &hn["ce_".len()..];
+
+            if attributes.contains(&name) {
+                visitor = visitor.set_attribute(
+                    name,
+                    MessageAttributeValue::String(String::from_utf8(hv).map_err(|e| {
+                        cloudevents::message::Error::Other {
+                            source: Box::new(e),
+                        }
+                    })?),
+                )?
+            } else {
+                visitor = visitor.set_extension(
+                    name,
+                    MessageAttributeValue::String(String::from_utf8(hv).map_err(|e| {
+                        cloudevents::message::Error::Other {
+                            source: Box::new(e),
+                        }
+                    })?),
+                )?
+            }
+        }
+
+        if self.payload != None {
+            visitor.end_with_data(self.payload.unwrap())
+        } else {
+            visitor.end()
+        }
+    }
+}
+
+impl StructuredDeserializer for ConsumerMessageDeserializer {
+    fn deserialize_structured<R: Sized, V: StructuredSerializer<R>>(self, visitor: V) -> Result<R> {
+        visitor.set_structured_event(self.payload.unwrap())
+    }
+}
+
+impl MessageDeserializer for ConsumerMessageDeserializer {
+    fn encoding(&self) -> Encoding {
+        match (
+            self.headers
+                .get("content-type")
+                .map(|s| String::from_utf8(s.to_vec()).ok())
+                .flatten()
+                .map(|s| s.starts_with(headers::CLOUDEVENTS_JSON_HEADER))
+                .unwrap_or(false),
+            self.headers.get(headers::SPEC_VERSION_HEADER),
+        ) {
+            (true, _) => Encoding::STRUCTURED,
+            (_, Some(_)) => Encoding::BINARY,
+            _ => Encoding::UNKNOWN,
+        }
+    }
+}
+
+pub fn record_to_event(msg: &Message, version: headers::MqttVersion) -> Result<Event> {
+    match version {
+        headers::MqttVersion::V5 => {
+            BinaryDeserializer::into_event(ConsumerMessageDeserializer::new(msg)?)
+        }
+        headers::MqttVersion::V3_1 => {
+            StructuredDeserializer::into_event(ConsumerMessageDeserializer::new(msg)?)
+        }
+        headers::MqttVersion::V3_1_1 => {
+            StructuredDeserializer::into_event(ConsumerMessageDeserializer::new(msg)?)
+        }
+    }
```

Comment:
I may miss some knowledge about mqtt here, but why do we need the headers version? Also, why in V5 only Binary Deserializer is supported (aka only binary mode is supported), more than the general MessageDeserializer?

Also, can you infer the headers version from `Message` more than providing it manually as parameter?

Reply:
now, V5 should allow both structured & binary mode. V3 only structured mode.